geopandas/tools/sjoin.py

```diff
@@ -186,60 +285,61 @@ def find_contains(a1, a2):
             columns={"_key_left": "_key_right", "_key_right": "_key_left"}
         )
 
-    if how == "inner":
-        result = result.set_index("_key_left")
-        joined = (
-            left_df.merge(result, left_index=True, right_index=True)
-            .merge(
-                right_df.drop(right_df.geometry.name, axis=1),
-                left_on="_key_right",
-                right_index=True,
-                suffixes=("_%s" % lsuffix, "_%s" % rsuffix),
-            )
-            .set_index(index_left)
-            .drop(["_key_right"], axis=1)
-        )
-        if isinstance(index_left, list):
-            joined.index.names = left_index_name
-        else:
-            joined.index.name = left_index_name
-
-    elif how == "left":
-        result = result.set_index("_key_left")
-        joined = (
-            left_df.merge(result, left_index=True, right_index=True, how="left")
-            .merge(
-                right_df.drop(right_df.geometry.name, axis=1),
-                how="left",
-                left_on="_key_right",
-                right_index=True,
-                suffixes=("_%s" % lsuffix, "_%s" % rsuffix),
-            )
-            .set_index(index_left)
-            .drop(["_key_right"], axis=1)
-        )
-        if isinstance(index_left, list):
-            joined.index.names = left_index_name
-        else:
-            joined.index.name = left_index_name
+    return result
 
-    else:  # how == 'right':
-        joined = (
-            left_df.drop(left_df.geometry.name, axis=1)
-            .merge(
-                result.merge(
-                    right_df, left_on="_key_right", right_index=True, how="right"
-                ),
-                left_index=True,
-                right_on="_key_left",
-                how="right",
+def _sjoin_nearest(left_df,
+    right_df,
+    tolerance,
+    nearest_distances,
+    tree_idx_r,
+    tree_idx_l,
+    ):
+    tree_idx = tree_idx_r or right_df.sindex
+    
+    if tolerance is None:
+        # pick a sensible tolerance
+        rel_tol = int(DEFAULT_NEAREST_REL_TOL * len(right_df) + DEFAULT_NEAREST_ABS_TOL)
```

Comment:
I tried to come up with some "default" `tolerance` to use if the user does not bother to input it. Not sure if this is what we want to do or not. I think it's important to have some type of warning because setting `tolerance` too small can lead to unexpected results but too big can lead to very poor performance.